mds/src/karatsuba_convolution.rs

```diff
@@ -0,0 +1,362 @@
+//!                  v_1(x) = v(x) mod x^{N/2} + 1
+//!                  u_0(x) = u(x) mod x^{N/2} - 1
+//!                  u_1(x) = u(x) mod x^{N/2} + 1
+//!
+//! Now w_0 is the convolution of v_0 and u_0 which we can compute
+//! recursively.  For w_1 we compute the signed convolution
+//! v_1(x)u_1(x) mod x^{N/2} + 1 using Karatsuba.
+//!
+//! There are 2 possible approaches to applying Karatsuba which mirror
+//! the DIT vs DIF approaches to FFT's.
+//!
+//! Option 1: left/right decomposition.
+//! Write v = (v_l, v_r) so that v(x) = (v_l(x) + x^{N/2}v_r(x)).
+//! Then v(x)u(x) mod x^N + 1
+//!      = (v_l(x)u_l(x) - v_r(x)u_r(x))
+//!        + x^{N/2}((v_l(x) + v_r(x))(u_l(x) + u_r(x))
+//!                   - (v_l(x)u_l(x) + v_r(x)u_r(x)))   (mod X^N + 1)
+//!
+//! As v_l(x), v_r(x), u_l(x), u_r(x) all have degree < N/2 no product
+//! will have degree > N - 1.  The only place we need to deal with the
+//! mod operation is after the multipication by x^{N/2} and this is
+//! easy to do.  Thus this reduces the problem to 3 polynomial
+//! multiplications of size N/2 and we can use standard Karatsuba
+//! for this.
+//!
+//! Option 2: even/odd decomposition.
+//! Define the even v_e and odd v_o parts so that v(x) = (v_e(x^2) + xv_o(x^2)).
+//! Then v(x)u(x)
+//!    = (v_e(x^2)u_e(x^2) + x^2 v_o(x^2)u_o(x^2))
+//!      + x ((v_e(x^2) + v_o(x^2))(u_e(x^2) + u_o(x^2))
+//!            - (v_e(x^2)u_e(x^2) + v_o(x^2)u_o(x^2)))
+//! This reduces the problem to 3 signed convolutions of size N/2 which
+//! can be computed recursively.
+//!
+//! Option 2 seems to involve less total operations and so should be
+//! faster hence it is the one implmented below.  The main issue is
+//! that it requires quite a bit of data manipulation (i.e. splitting
+//! vectors into odd and even parts and recombining afterwards).
+//! It might be beneficial to find a way to cut down on this.
+//!
+//! Of course, for small sizes we just explicitly write out the O(n^2)
+//! approach.
+
+use core::ops::{Add, AddAssign, Neg, ShrAssign, Sub, SubAssign};
+
+/// This trait collects the operations needed by `Convolve` below.
+///
+/// TODO: Think of a better name for this.
+pub trait RngElt:
+    Add<Output = Self>
+    + AddAssign
+    + Copy
+    + Default
+    + Neg<Output = Self>
+    + ShrAssign<u32>
+    + Sub<Output = Self>
+    + SubAssign
+{
+}
+
+impl RngElt for i64 {}
+impl RngElt for i128 {}
+
+/// Template function to perform convolution of vectors.
+///
+/// Roughly speaking, for a convolution of size `N`, it should be
+/// possible to add `N` elements of type `T` without overflowing, and
+/// similarly for `U`. Then multiplication via `Self::mul` should
+/// produce an element of type `V` which will not overflow after about
+/// `N` additions (this is an over-estimate).
+///
+/// NB: Note that the convolution code does some `ShrAssign`s after
+/// calling `mul`, so if `mul` does an intermediate/partial reduction,
```

Comment:
Maybe comment that we use specialized partial reductions so we can still divide by 2 afterwards?